File: JavaScript/ChaserGhost.js
import { Ghost } from "./Ghost"

export class ChaserGhost extends Ghost
{
	constructor()
	{
		this.color = "#8888FF";
	}

	special(map, pacman, ghosts, mapSize)
	{
		if (this.lineOfSight(map, pacman, ghosts, mapSize) && !pacman.getPower())
		{
			this.setTimer(2 * 60);
			this.hasLineOfSight = true;
		}
		else
			this.hasLineOfSight = false;

		if (this.timer && !this.firstTarget)
		{
			this.hasTarget = true;
			this.target = pacman.getPos();
			this.speed = this.defSpeed + 2 + this.hasLineOfSight * 3;
		}
		else
			this.speed = this.defSpeed;
	}

	getType()
	{
		return 0;
	}
}

File: C++/ChaserGhost.cpp
#include "ChaserGhost.h"

ChaserGhost::ChaserGhost()
{
	color = sf::Color(0x8888FFFF);
}

void ChaserGhost::special(std::vector<std::vector<char>>& map, std::vector<Pacman>& pacmen, std::vector<Ghost*>& ghosts, sf::Vector2i mapSize)
{
	hasLineOfSight = false;

	for (int i = 0; i < pacmen.size(); i++)
	{
		if (lineOfSight(map, pacmen[i], ghosts, mapSize) && !pacmen[i].getPower() && isFree)
		{
			setTimer(2 * 60);
			targetPacman = i;
			hasLineOfSight = true;
		}

		if (timer)
		{
			firstTarget = false;
			hasTarget = true;
			if (targetPacman != -1)
			target = pacmen[targetPacman].getPos();
			speed = defSpeed + 2 + hasLineOfSight * 3;
		}
		else
		{
			speed = defSpeed;
			targetPacman = -1;
		}
	}
}

void ChaserGhost::assignPacman(int index)
{
}
